pkg/server/commands/reverseexpand/reverse_expand_weighted.go

```diff
@@ -0,0 +1,519 @@
+package reverseexpand
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"go.opentelemetry.io/otel/trace"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	weightedGraph "github.com/openfga/language/pkg/go/graph"
+
+	"github.com/openfga/openfga/internal/concurrency"
+	"github.com/openfga/openfga/internal/condition"
+	"github.com/openfga/openfga/internal/condition/eval"
+	"github.com/openfga/openfga/internal/utils"
+	"github.com/openfga/openfga/internal/validation"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/telemetry"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+type typeRelEntry struct {
+	typeRel string // e.g. "organization#admin"
+
+	// Only present for userset relations. Will be the userset relation string itself.
+	// For `rel admin: [team#member]`, usersetRelation is "member"
+	usersetRelation string
+}
+
+// relationStack represents the path of queryable relationships encountered on the way to a terminal type.
+// As reverseExpand traverses from a requested type#rel to its leaf nodes, it pushes to this stack.
+// Each entry is a `typeRelEntry` struct, which contains the `type#relation`
+// along with metadata:
+//   - `usersetRelation`: To handle transitions through usersets (e.g. `[team#member]`).
+//
+// After reaching a leaf, this stack is consumed by the `queryForTuples` function to build the precise chain of
+// database queries needed to find the resulting objects.
+// To avoid races, every leaf node receives its own copy of the stack.
+type relationStack []typeRelEntry
```

Comment:
Is this just for convenience? https://github.com/emirpasic/gods?tab=readme-ov-file#arraystack